[PySpark] Add functions covar_pop, covar_samp, call_functions, endswith, startswith, exp, factorial, log2, ln, degrees, radians, atan, atan2, tan, round, bround

binste opened the pull request:
I'll just keep creating new small PRs (e.g. #14347). Let me know if you'd prefer less frequent but larger ones.

binste committed:
Add call_functions function

binste committed:
Add covar_pop function

binste committed:
Add covar_samp

binste committed:
Add endswith function

binste committed:
Add startswith function

binste committed:
Add exp function

binste committed:
Add factorial function

binste committed:
Add log2 function

binste committed:
Add ln function

binste committed:
Format code

binste committed:
Add degrees function

binste committed:
Add radians function

binste committed:
Add atan and atan2 functions

binste committed:
Add tan function

binste committed:
Add round function

binste committed:
Add bround function

binste committed:
Format code

Tishj commented on tools/pythonpkg/tests/fast/spark/test_spark_group_by.py:
```diff
@@ -116,6 +116,22 @@ def test_group_by(self, spark):
             == "[Row(department='Finance', sum_salary=351000, avg_salary=87750.0, sum_bonus=81000, max_bonus=24000), Row(department='Sales', sum_salary=257000, avg_salary=85666.66666666667, sum_bonus=53000, max_bonus=23000)]"
         )
 
+        df = spark.createDataFrame(
```
If possible, try to create separate methods for tests, if this were to fail in the future it would be difficult to debug as that work to separate out the relevant parts would need to be done afterwards, but for now let's leave this as is :+1:

binste replied:
I got the impression from some of the existing tests that the preference is towards not separating it too much but totally agree, I also prefer smaller self-contained tests. Thanks for the feedback, will do!

Tishj reviewed (approved):
Thanks for the PR!

Bigger PRs are fine if the implementations are trivial such as these ones 👍

Mytherin commented:
Thanks for the PR! Could you just resolve the merge conflicts?

binste committed:
Somehow got an error that ends_with does not exist. Using suffix now

binste committed:
Format code

binste commented:
Sorry for creating multiple PRs which modify the same code areas. I've now rebased on `feature` and fixed the merge conflicts. I'll resolve the conflicts in #14458 once this PR is merged.

In general, should I start new features always on top of `feature` and bug fixes on top of `main`?

Mytherin commented:
Thanks!

> In general, should I start new features always on top of feature and bug fixes on top of main?

Yes, new features should go into `feature` if there is a feature branch (we will merge it into main soon).
